lib/Transforms/HC/SelectAcceleratorCode/SelectAcceleratorCode.cpp

```diff
@@ -0,0 +1,175 @@
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/DiagnosticInfo.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <algorithm>
+#include <unordered_set>
+
+using namespace llvm;
+using namespace std;
+
+namespace
+{
+    class SelectAcceleratorCode : public ModulePass {
+        unordered_set<Function*> hcCallees_;
+
+        void findAllHCCallees_(const Function &f, Module &M)
+        {
+            for (auto&& x : f) {
+                for (auto&& y : x) {
+                    if (y.getOpcode() == Instruction::Call) {
+                        auto g = cast<CallInst>(y).getCalledFunction();
+                        if (g) {
+                            auto t =
+                                hcCallees_.insert(M.getFunction(g->getName()));
+                            if (t.second) findAllHCCallees_(*g, M);
+                        }
+                    }
+                }
+            }
+        }
+
+        template<typename T>
+        static
+        void erase_(T& x)
+        {
+            x.dropAllReferences();
+            x.replaceAllUsesWith(UndefValue::get(x.getType()));
+            x.eraseFromParent();
+        }
+
+        template<typename F, typename G, typename P>
+        bool eraseIf_(F f, G l, P p) const
+        {
+            bool r = false;
+
+            auto it = f();
+            while (it != l()) {
+                it->removeDeadConstantUsers();
+                if (p(*it)) {
+                    erase_(*it);
+                    r = true;
+                    it = f();
+                }
+                else ++it;
+            }
+
+            return r;
+        }
+
+        bool eraseNonHCFunctions_(Module &M) const
+        {
+            return eraseIf_(
+                [&]() { return M.begin(); },
+                [&]() { return M.end(); },
+                [&, this](const Function& x) {
+                    return hcCallees_.count(M.getFunction(x.getName())) == 0;
+                });
+        }
+
+        bool eraseDeadGlobals_(Module &M) const
+        {
+            return eraseIf_(
+                [&]() { return M.global_begin(); },
+                [&]() { return M.global_end(); },
+                [](const Constant& x) { return !x.isConstantUsed(); });
+        }
+
+        bool eraseDeadAliases_(Module &M)
+        {
+            return eraseIf_(
+                [&]() { return M.alias_begin(); },
+                [&]() { return M.alias_end(); },
+                [](const Constant& x) { return !x.isConstantUsed(); });
+        }
+
+        static
+        bool alwaysInline_(Function& x)
+        {
+            if (!x.hasFnAttribute(Attribute::AlwaysInline)) {
+                if (x.hasFnAttribute(Attribute::NoInline)) {
+                    x.removeFnAttr(Attribute::NoInline);
+                }
+                x.addFnAttr(Attribute::AlwaysInline);
```

Comment:
this is where you all functions visited by HC kernels/functions with `alwaysinline` attribute.